### PR TITLE
(feat) implement EIP-1153 - transient storage

### PR DESCRIPTION
**Pull request by NazariiDenha**
## 1. Purpose or design rationale of this PR

https://eips.ethereum.org/EIPS/eip-1153
https://github.com/ethereum/execution-specs/blob/master/network-upgrades/mainnet-upgrades/cancun.md
https://www.notion.so/scrollzkp/Add-tload-tstore-opcodes-to-l2geth-8a54dbbcdb894ebd8e6d9c4d46b6de9a?pvs=4

## 2. PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [ ] build: Changes that affect the build system or external dependencies (example scopes: yarn, eslint, typescript)
- [ ] ci: Changes to our CI configuration files and scripts (example scopes: vercel, github, cypress)
- [ ] docs: Documentation-only changes
- [x] feat: A new feature
- [ ] fix: A bug fix
- [ ] perf: A code change that improves performance
- [ ] refactor: A code change that doesn't fix a bug, or add a feature, or improves performance
- [ ] style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
- [ ] test: Adding missing tests or correcting existing tests

## 3. Deployment tag versioning

Has the version in `params/version.go` been updated?

- [ ] This PR doesn't involve a new deployment, git tag, docker image tag, and it doesn't affect traces
- [x] Yes

## 4. Breaking change label

Does this PR have the `breaking-change` label?

- [ ] This PR is not a breaking change
- [ ] Yes

### CODE REVIEW

**Comment by 0xmountaintop** [comment]
This PR looks good, but personally I will prefer breaking https://github.com/scroll-tech/go-ethereum/pull/701/commits/34a13ff7bba7bd46b80579e195a19b0dcd34abbe into
+ cherry-pick https://github.com/ethereum/go-ethereum/pull/27613
+ enable 1153
+ other fixes

File: core/vm/jump_table.go [hunk 1]
```diff
@@ -70,7 +70,7 @@ type JumpTable [256]*operation
 // contantinople, istanbul, petersburg, berlin, london, shanghai, curie, and descartes instructions.
 func newDescartesInstructionSet() JumpTable {
 	instructionSet := newCurieInstructionSet()
-
+	enable1153(&instructionSet)
```
Comment (Thegaram) [inline core/vm/jump_table.go]: Let's keep this PR open for now. We have not decided whether this should go into Curie or Descartes.

**Review by Thegaram — CHANGES_REQUESTED** [review]

File: core/state/statedb.go [hunk 2]
```diff
@@ -1018,34 +1054,44 @@ func (s *StateDB) Commit(deleteEmptyObjects bool) (common.Hash, error) {
...
-	for _, el := range list {
-		s.AddAddressToAccessList(el.Address)
-		for _, key := range el.StorageKeys {
-			s.AddSlotToAccessList(el.Address, key)
+// Potential EIPs:
+// - Reset access list (Berlin)
+// - Add coinbase to access list (EIP-3651)
+// - Reset transient storage (EIP-1153)
+func (s *StateDB) Prepare(rules params.Rules, sender, coinbase common.Address, dst *common.Address, precompiles []common.Address, list types.AccessList) {
+	if rules.IsBerlin {
```
Comment (Thegaram) [inline core/state/statedb.go:1071]: This part is not the same as in upstream: https://github.com/ethereum/go-ethereum/blob/master/core/state/statedb.go#L1277

File: core/state/transient_storage.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+type transientStorage map[common.Address]Storage
+
+// newTransientStorage creates a new instance of a transientStorage.
+func newTransientStorage() transientStorage {
+	return make(transientStorage)
+}
+
+// Set sets the transient-storage `value` for `key` at the given `addr`.
+func (t transientStorage) Set(addr common.Address, key, value common.Hash) {
+	if _, ok := t[addr]; !ok {
```
Comment (Thegaram) [inline core/state/transient_storage.go:33]: Also: https://github.com/ethereum/go-ethereum/blob/master/core/state/transient_storage.go#L37

File: core/vm/interface.go [hunk 4]
```diff
@@ -54,6 +54,8 @@ type StateDB interface {
 	GetProof(addr common.Address) ([][]byte, error)
 	GetProofByHash(addrHash common.Hash) ([][]byte, error)
 	GetStorageProof(a common.Address, key common.Hash) ([][]byte, error)
+	GetTransientState(addr common.Address, key common.Hash) common.Hash
```
Comment (Thegaram) [inline core/vm/interface.go:57]:
```suggestion

	GetTransientState(addr common.Address, key common.Hash) common.Hash
```

**Comment by NazariiDenha** [comment]
closing in favour of https://github.com/scroll-tech/go-ethereum/pull/721

**Pull request closed without merging**